### PR TITLE
WIP: Get property type

### PR DESCRIPTION
**Pull request by brawner**
These are new changes to help port rqt_publisher. It adds a couple of helper functions to topic_helpers.py and makes a fix to get_slot_type. Two helper functions translate between property names and slot names and vice versa. I also added a get_property_type function that provides similar functionality to get_slot_type just using the property_path. I've modified logic in message_tree_model.py to reflect the new get_property_type function.

New commit: https://github.com/ros-visualization/rqt/commit/b13f83277fc67701f3ce297b6a96cab99a1a793b

### CODE REVIEW

File: rqt_py_common/src/rqt_py_common/topic_helpers.py [hunk 1]
```diff
@@ -244,10 +244,29 @@ def _get_field_type(topic_names_and_types, target):  # noqa: C901
...
+    unpublished topics and with empty arrays.
+
+    :param message_class: message class type, ``type``
+    :param property_path: path to the property inside the message class, ``str``,
+        i.e. 'header/seq'
+    :returns: property_type, is_array
+    """
+    is_array = False
+    properties = [f for f in property_path.split('/') if f]
+    slot_path = '/'.join(['_' + p for p in properties])
```
Comment (mlautman) [inline rqt_py_common/src/rqt_py_common/topic_helpers.py:266]: are we still using underscores? the new changes to rosidl_python should make this cleaner https://github.com/ros2/rosidl_python/pull/19

File: rqt_py_common/src/rqt_py_common/topic_helpers.py [hunk 2]
```diff
@@ -256,21 +275,94 @@ def get_slot_type(message_class, slot_path):
...
+        # we want to return the type of the array without setting is_array to True
+        is_array_component = False
+        array_index = field_name.find('[')
+        if array_index >= 0:
+            field_name = field_name[:array_index]
+            is_array_component = True
+
+        # Get the slot type
+        slot_index = message_class.__slots__.index(field_name)
+        slot_class_name = message_class._slot_types[slot_index]
```
Comment (mlautman) [inline rqt_py_common/src/rqt_py_common/topic_helpers.py:296]: ros2/rosidl_python#19

File: rqt_py_common/src/rqt_py_common/topic_helpers.py [hunk 3]
```diff
@@ -256,21 +275,94 @@ def get_slot_type(message_class, slot_path):
...
+        raise ValueError('Invalid property name, use the property name without array indices')
+    if not hasattr(message_class, property_name):
+        raise ValueError('Message class "{}" does not have a property name "{}"'.format(
+            message_class, property_name))
+    if not isinstance(getattr(message_class, property_name), property):
+        raise ValueError('The field "{}" in message class "{}" is not a property'.format(
+            property_name, message_class))
+
+    # slots start with an underscore
+    slot_name = '_' + property_name
```
Comment (mlautman) [inline rqt_py_common/src/rqt_py_common/topic_helpers.py:362]: ros2/rosidl_python#19

File: rqt_py_common/src/rqt_py_common/topic_helpers.py [hunk 4]
```diff
@@ -256,21 +275,94 @@ def get_slot_type(message_class, slot_path):
...
+
+    if not hasattr(message_class, slot_name):
+        raise ValueError('Message class "{}" does not have a slot name "{}"'.format(
+            message_class, slot_name))
+
+    if isinstance(getattr(message_class, slot_name), property):
+        return slot_name
+
+    # slots start with an underscore
+    property_name = slot_name[1:]
```
Comment (mlautman) [inline rqt_py_common/src/rqt_py_common/topic_helpers.py:332]: ros2/rosidl_python#19

**Review by mlautman — CHANGES_REQUESTED** [review]

File: rqt_py_common/src/rqt_py_common/topic_helpers.py (hunk 1 above)
Comment (brawner) [reply rqt_py_common/src/rqt_py_common/topic_helpers.py:266]: I'll update based on that new PR

Comment (mlautman) [reply rqt_py_common/src/rqt_py_common/topic_helpers.py:266]: Also, take a look at #162 for some api changes to the helper functions

**Comment by brawner** [comment]
This was subsumed by #162

**Pull request closed without merging**